controllers/cps_controller_helper.go

```diff
@@ -0,0 +1,217 @@
+	"sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+var setupLog = ctrl.Log.WithName("setup")
+
+func (r *GatekeeperReconciler) addConstraintController(mainCtx context.Context,
+	gatekeeper *operatorv1alpha1.Gatekeeper,
+) error {
+	ready, err := checkCPScontrollerAvailable(mainCtx, gatekeeper, r.DynamicClient)
+	if err != nil {
+		return err
+	}
+
+	// auditFromCache is not Automatic, Delete existing ConstraintPodStatus Controller
+	if !ready {
+		if r.childCtx != nil {
+			r.DeleteCPSController()
+		}
+
+		return nil
+	}
+
+	if r.isCpsCtlInstalled {
+		return nil
+	}
+
+	r.childCtx, r.childCtxCancel = context.WithCancel(mainCtx)
+
+	cpsMgr, err := ctrl.NewManager(r.Cfg, ctrl.Options{
+		Scheme: r.Scheme,
+		Metrics: server.Options{
+			BindAddress: ":8083",
+		},
+		HealthProbeBindAddress: ":8084",
+		LeaderElection:         r.EnableLeaderElection,
+		LeaderElectionID:       "5ff985ccc.constraintstatuspod.gatekeeper.sh",
+		Cache: cacheRuntime.Options{
+			ByObject: map[client.Object]cacheRuntime.ByObject{
+				&gkv1beta1.ConstraintPodStatus{}: {
+					Transform: func(obj interface{}) (interface{}, error) {
+						constraintStatus := obj.(*gkv1beta1.ConstraintPodStatus)
+						// Only cache fields that are utilized by the controllers.
+						guttedObj := &gkv1beta1.ConstraintPodStatus{
+							TypeMeta: constraintStatus.TypeMeta,
+							ObjectMeta: metav1.ObjectMeta{
+								Name:      constraintStatus.Name,
+								Labels:    constraintStatus.Labels,
+								Namespace: constraintStatus.Namespace,
+							},
+							Status: gkv1beta1.ConstraintPodStatusStatus{
+								ObservedGeneration: constraintStatus.Status.ObservedGeneration,
+								Operations:         constraintStatus.Status.Operations,
+							},
+						}
+
+						return guttedObj, nil
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		setupLog.Error(err, "Failed to setup NewManager for ConstraintPodstatus contoller")
+
+		return err
+	}
+
+	// Use another go routine for gatekeeper controller
+	go func() {
+		err = r.addConstraintStatusManager(cpsMgr)
+		if err != nil {
+			r.isCpsCtlInstalled = false
+
+			setupLog.Error(err, "unable to start ConstraintPodStatus manager")
+		}
+	}()
+
+	return nil
+}
+
+func (r *GatekeeperReconciler) addConstraintStatusManager(mgr manager.Manager,
+) error {
+	if err := (&ConstraintStatusReconciler{
+		Scheme:               r.Scheme,
+		Client:               mgr.GetClient(),
+		ClientSet:            kubernetes.NewForConfigOrDie(mgr.GetConfig()),
+		DynamicClient:        r.DynamicClient,
+		Log:                  ctrl.Log.WithName("controllers").WithName("ConstraintPodStatus"),
+		Namespace:            r.Namespace,
+		ConstraintToSyncOnly: map[string][]gkv1alpha1.SyncOnlyEntry{},
+	}).SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", "ConstraintPodStatus")
+
+		return err
+	}
+
+	if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
```

Comment:
These health endpoints aren't used since Kubernetes can just be configured to use one HTTP endpoint. In the governance-policy-framework-addon, I wrote an HTTP endpoint that combined both health checks. You could do that or have the GatekeeperReconciler health check include a custom health check.

Reply:
I have been thinking... If the child ctx closes, do we have to send a liveness probe 500 signal? We close the child ctx when `Automatic` is removed without closing the gatekeeper controller. Moreover, It restarts the constraintPodStatus controller when it has an error then `child ctx` closes just like it acts bad heath check, liveness probe 500.

Reply:
Also, I tried to add proxyHealth there is a limitation. First, I created `child ctx` in `gatekeeper controller`. so I can create a variable sync with `child ctx`. but l think looks too complicated unnecessarily. We already have a very simple way. Install `contraintPodstatus CRD`, check `Automation`  in `contraintPodstatus` pod, If `Automation` is set, proceed rest of the reconcile function.  But as I mentioned at the first comment, I think we don't need to add health check for this controller

Reply:
Okay, just remove them since they are useless.